# Loading

In [1]:
import pandas as pd

In [2]:
y = pd.read_csv('df_stem.csv')

In [3]:
y = y.sentiment

In [4]:
y.head()

0    1
1    1
2    1
3    0
4    1
Name: sentiment, dtype: int64

In [5]:
x = pd.read_csv('df_vector.csv')

In [6]:
x.head()

Unnamed: 0         0         1         2         3         4         5   
0           0 -0.251147  0.561305  0.126222 -0.094850  0.114038 -0.859551  \
1           1 -0.351663  0.121878 -0.120232 -0.553368  0.182102 -1.230392   
2           2 -0.552294  0.300187 -0.094526 -0.310396 -0.018994 -0.820702   
3           3 -0.233144  0.494998  0.041341 -0.211720 -0.121327 -0.502161   
4           4 -0.376380  0.266668 -0.131451 -0.053470  0.472519 -1.122471   

          6         7         8  ...        90        91        92        93   
0 -0.240192  0.361602 -0.432211  ...  0.321737  0.545179  0.062861  0.241012  \
1  0.201216  0.558062 -0.544723  ...  0.239608  0.648976  0.118813  0.202644   
2  0.353900  0.716587 -0.478776  ...  0.316577  0.510552  0.079388  0.559151   
3  0.055911  0.394001 -0.394661  ...  0.094764  0.917627  0.364555  0.159884   
4  0.230775  0.287217 -0.643272  ...  0.239622  0.739376  0.068496  0.188084   

         94        95        96        97        98        99  
0 -0.001750  0.349589  0.236443 -0.480595  0.375254 -0.430951  
1 -0.037746  0.672800  0.460015 -0.634437  0.277175 -0.146046  
2 -0.047210  0.351930  0.533410 -0.288712  0.183992 -0.352972  
3 -0.202905  0.805818  0.485097 -0.742107  0.061896 -0.303398  
4  0.096164  0.574591  0.168376 -0.363870  0.606393 -0.529876  

[5 rows x 101 columns]

In [7]:
x = x.iloc[:, 1:]

## Training and Testing

In [12]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [11]:
import keras

In [33]:
def train_and_test(layers, epochs, x, y):
    x_train, x_test, y_train, y_test = train_test_split(
        x, y,
        test_size = 0.2
    )
    
    model = keras.models.Sequential()
    for units in layers:
        model.add(
            keras.layers.Dense(
                units = units,
                kernel_initializer = 'he_uniform',
                activation = 'relu'
            )
        )
    
    model.add(
        keras.layers.Dense(
            units = 1,
            kernel_initializer = 'glorot_uniform',
            activation = 'sigmoid',
        )
    )
    
    model.compile(
        optimizer = 'adam',
        loss = 'binary_crossentropy',
        metrics = ['accuracy']
    )
    
    return model.fit(
        x_train,
        y_train,
        validation_split = 0.33,
        batch_size = 100,
        epochs = epochs
    )

### ANN with 1 layer of 50 neurons (without scaling)

In [34]:
train_and_test([50], 100, x, y)

Epoch 1/100
268/268 [==============================] - 1s 3ms/step - loss: 0.4679 - accuracy: 0.7839 - val_loss: 0.3909 - val_accuracy: 0.8281
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3603 - accuracy: 0.8456 - val_loss: 0.3608 - val_accuracy: 0.8436
Epoch 3/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3476 - accuracy: 0.8507 - val_loss: 0.3494 - val_accuracy: 0.8496
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3409 - accuracy: 0.8544 - val_loss: 0.3449 - val_accuracy: 0.8549
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3374 - accuracy: 0.8559 - val_loss: 0.3451 - val_accuracy: 0.8491
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3337 - accuracy: 0.8568 - val_loss: 0.3391 - val_accuracy: 0.8551
Epoch 7/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3318 - accuracy: 0.8581 - val_loss: 0.3384 - val_accuracy: 0.8549

89.25% training accuracy with 85.68% validation accuracy

### ANN with 2 layers of 50 and 10 neurons (without scaling)

In [35]:
train_and_test([50, 10], 100, x, y)

Epoch 1/100
268/268 [==============================] - 1s 3ms/step - loss: 0.4467 - accuracy: 0.8043 - val_loss: 0.3596 - val_accuracy: 0.8429
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3473 - accuracy: 0.8505 - val_loss: 0.3408 - val_accuracy: 0.8544
Epoch 3/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3407 - accuracy: 0.8543 - val_loss: 0.3434 - val_accuracy: 0.8527
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3347 - accuracy: 0.8576 - val_loss: 0.3342 - val_accuracy: 0.8556
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3314 - accuracy: 0.8569 - val_loss: 0.3331 - val_accuracy: 0.8577
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3304 - accuracy: 0.8567 - val_loss: 0.3316 - val_accuracy: 0.8587
Epoch 7/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3291 - accuracy: 0.8575 - val_loss: 0.3345 - val_accuracy: 0.8551

89.73% training accuracy with 85.45% validation accuracy

### ANN with 2 layers of 60 and 20 neurons (without scaling)

In [36]:
train_and_test([60, 20], 100, x, y)

Epoch 1/100
268/268 [==============================] - 1s 3ms/step - loss: 0.4235 - accuracy: 0.8075 - val_loss: 0.3648 - val_accuracy: 0.8399
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3420 - accuracy: 0.8523 - val_loss: 0.3520 - val_accuracy: 0.8501
Epoch 3/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3330 - accuracy: 0.8571 - val_loss: 0.3510 - val_accuracy: 0.8485
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3291 - accuracy: 0.8585 - val_loss: 0.3450 - val_accuracy: 0.8510
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3245 - accuracy: 0.8606 - val_loss: 0.3424 - val_accuracy: 0.8530
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3202 - accuracy: 0.8626 - val_loss: 0.3439 - val_accuracy: 0.8517
Epoch 7/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3204 - accuracy: 0.8614 - val_loss: 0.3440 - val_accuracy: 0.8508

91.23% training accuracy with 84.36% validation accuracy

- slightly overfitting
- should add dropout

### ANN with 2 layers of 60 and 20 neurons (with Standard scaling)

In [42]:
from sklearn.preprocessing import StandardScaler

In [43]:
x_scaled = pd.DataFrame(StandardScaler().fit_transform(x), columns=x.columns)

In [44]:
x_scaled.describe()

0             1             2             3             4   
count  5.000000e+04  5.000000e+04  5.000000e+04  5.000000e+04  5.000000e+04  \
mean  -3.173284e-16  3.925038e-16 -9.791279e-17  8.327561e-17  2.842171e-18   
std    1.000010e+00  1.000010e+00  1.000010e+00  1.000010e+00  1.000010e+00   
min   -6.202139e+00 -1.486309e+01 -6.757726e+00 -7.300654e+00 -7.937158e+00   
25%   -6.225139e-01 -6.173718e-01 -6.262052e-01 -6.319055e-01 -6.305686e-01   
50%    3.536865e-02  1.606233e-02  4.143598e-02  2.455183e-02  5.129331e-02   
75%    6.604330e-01  6.423465e-01  6.776879e-01  6.542170e-01  6.830048e-01   
max    5.068933e+00  4.830957e+00  4.701224e+00  4.675281e+00  5.263099e+00   

                  5             6             7             8             9   
count  5.000000e+04  5.000000e+04  5.000000e+04  5.000000e+04  5.000000e+04  \
mean  -5.124434e-16  5.172751e-17 -3.666401e-17  2.529532e-17 -8.014922e-17   
std    1.000010e+00  1.000010e+00  1.000010e+00  1.000010e+00  1.000010e+00   
min   -4.778139e+00 -6.168499e+00 -6.339533e+00 -7.330723e+00 -7.146061e+00   
25%   -6.539860e-01 -6.427274e-01 -6.326161e-01 -6.198444e-01 -6.096701e-01   
50%    1.163906e-02 -2.702323e-02 -5.582847e-03  1.585602e-02  9.346582e-02   
75%    6.624647e-01  6.158727e-01  6.231269e-01  6.372821e-01  7.004243e-01   
max    5.020806e+00  6.138458e+00  5.728607e+00  5.486503e+00  4.661029e+00   

       ...            90            91            92            93   
count  ...  5.000000e+04  5.000000e+04  5.000000e+04  5.000000e+04  \
mean   ...  2.636824e-16  5.329071e-16 -2.856382e-17  1.239187e-16   
std    ...  1.000010e+00  1.000010e+00  1.000010e+00  1.000010e+00   
min    ... -4.783539e+00 -7.973312e+00 -4.788789e+00 -7.246421e+00   
25%    ... -6.435937e-01 -6.397675e-01 -6.614029e-01 -6.346644e-01   
50%    ... -4.012027e-02  1.235018e-03 -4.511427e-02 -4.334273e-04   
75%    ...  6.120204e-01  6.385443e-01  6.103876e-01  6.315670e-01   
max    ...  6.400360e+00  5.735487e+00  6.577711e+00  6.951266e+00   

                 94            95            96            97            98   
count  5.000000e+04  5.000000e+04  5.000000e+04  5.000000e+04  5.000000e+04  \
mean  -2.842171e-17 -6.414780e-16  3.038281e-16 -2.927436e-16 -4.433787e-17   
std    1.000010e+00  1.000010e+00  1.000010e+00  1.000010e+00  1.000010e+00   
min   -5.562719e+00 -4.311398e+00 -5.486682e+00 -6.412036e+00 -5.199415e+00   
25%   -6.166865e-01 -6.801480e-01 -6.601366e-01 -6.117251e-01 -6.418793e-01   
50%    4.115596e-02 -6.526293e-02 -1.300968e-02  2.048285e-02  3.744644e-02   
75%    6.560017e-01  6.227306e-01  6.463611e-01  6.382716e-01  6.811852e-01   
max    5.399605e+00  5.530861e+00  6.506093e+00  6.276730e+00  4.186152e+00   

                 99  
count  5.000000e+04  
mean  -3.026202e-16  
std    1.000010e+00  
min   -7.108011e+00  
25%   -6.227626e-01  
50%    3.542545e-02  
75%    6.553488e-01  
max    6.961047e+00  

[8 rows x 100 columns]

In [45]:
train_and_test([60, 20], 100, x_scaled, y)

Epoch 1/100
268/268 [==============================] - 1s 2ms/step - loss: 0.4154 - accuracy: 0.8045 - val_loss: 0.3502 - val_accuracy: 0.8487
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3340 - accuracy: 0.8575 - val_loss: 0.3386 - val_accuracy: 0.8536
Epoch 3/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3202 - accuracy: 0.8654 - val_loss: 0.3377 - val_accuracy: 0.8545
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3118 - accuracy: 0.8677 - val_loss: 0.3316 - val_accuracy: 0.8577
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3057 - accuracy: 0.8704 - val_loss: 0.3286 - val_accuracy: 0.8585
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3000 - accuracy: 0.8725 - val_loss: 0.3351 - val_accuracy: 0.8555
Epoch 7/100
268/268 [==============================] - 1s 2ms/step - loss: 0.2953 - accuracy: 0.8750 - val_loss: 0.3303 - val_accuracy: 0.8577

96.39% training accuracy with 82.11% validation accuracy

- highly overfitting
- ADD DROPOUT
- try other scalers

### ANN with 2 layers of 60 and 20 neurons (with MinMax scaling)

In [46]:
from sklearn.preprocessing import MinMaxScaler

In [47]:
x_scaled = pd.DataFrame(MinMaxScaler().fit_transform(x), columns=x.columns)

In [48]:
x_scaled.describe()

0             1             2             3             4   
count  50000.000000  50000.000000  50000.000000  50000.000000  50000.000000  \
mean       0.550271      0.754700      0.589733      0.609610      0.601288   
std        0.088724      0.050777      0.087269      0.083502      0.075757   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.495039      0.723352      0.535086      0.556846      0.553519   
50%        0.553409      0.755515      0.593349      0.611660      0.605174   
75%        0.608866      0.787316      0.648874      0.664238      0.653030   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

                  5             6             7             8             9   
count  50000.000000  50000.000000  50000.000000  50000.000000  50000.000000  \
mean       0.487618      0.501220      0.525312      0.571943      0.605235   
std        0.102053      0.081256      0.082864      0.078021      0.084696   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.420877      0.448996      0.472891      0.523583      0.553599   
50%        0.488805      0.499025      0.524849      0.573180      0.613151   
75%        0.555223      0.551263      0.576946      0.621664      0.664557   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

       ...            90            91            92            93   
count  ...  50000.000000  50000.000000  50000.000000  50000.000000  \
mean   ...      0.427717      0.581620      0.421307      0.510394   
std    ...      0.089415      0.072947      0.087979      0.070435   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.370170      0.534952      0.363118      0.465693   
50%    ...      0.424129      0.581710      0.417338      0.510364   
75%    ...      0.482440      0.628199      0.475008      0.554878   
max    ...      1.000000      1.000000      1.000000      1.000000   

                 94            95            96            97            98   
count  50000.000000  50000.000000  50000.000000  50000.000000  50000.000000  \
mean       0.507440      0.438050      0.457499      0.505332      0.553980   
std        0.091222      0.101604      0.083384      0.078811      0.106548   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.451185      0.368945      0.402454      0.457122      0.485590   
50%        0.511194      0.431419      0.456414      0.506946      0.557970   
75%        0.567281      0.501321      0.511395      0.555634      0.626558   
max        1.000000      1.000000      1.000000      1.000000      1.000000   

                 99  
count  50000.000000  
mean       0.505223  
std        0.071079  
min        0.000000  
25%        0.460958  
50%        0.507741  
75%        0.551804  
max        1.000000  

[8 rows x 100 columns]

In [49]:
train_and_test([60, 20], 100, x_scaled, y)

Epoch 1/100
268/268 [==============================] - 1s 3ms/step - loss: 0.5706 - accuracy: 0.7226 - val_loss: 0.4439 - val_accuracy: 0.8082
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.4083 - accuracy: 0.8169 - val_loss: 0.3763 - val_accuracy: 0.8378
Epoch 3/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3790 - accuracy: 0.8338 - val_loss: 0.3601 - val_accuracy: 0.8474
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3680 - accuracy: 0.8390 - val_loss: 0.3597 - val_accuracy: 0.8453
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3601 - accuracy: 0.8426 - val_loss: 0.3495 - val_accuracy: 0.8503
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3606 - accuracy: 0.8418 - val_loss: 0.3485 - val_accuracy: 0.8505
Epoch 7/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3568 - accuracy: 0.8447 - val_loss: 0.3463 - val_accuracy: 0.8524

86.11% training accuracy with 85.65% validation accuracy

- accuracy dropped

In [51]:
def train_and_test_add_dropout(layers, dropouts, epochs, x, y):
    x_train, x_test, y_train, y_test = train_test_split(
        x, y,
        test_size = 0.2
    )
    
    model = keras.models.Sequential()
    for units, dropout in zip(layers, dropouts):
        model.add(
            keras.layers.Dropout(dropout)
        )
        model.add(
            keras.layers.Dense(
                units = units,
                kernel_initializer = 'he_uniform',
                activation = 'relu'
            )
        )
    
    model.add(
        keras.layers.Dense(
            units = 1,
            kernel_initializer = 'glorot_uniform',
            activation = 'sigmoid',
        )
    )
    
    model.compile(
        optimizer = 'adam',
        loss = 'binary_crossentropy',
        metrics = ['accuracy']
    )
    
    return model.fit(
        x_train,
        y_train,
        validation_split = 0.33,
        batch_size = 100,
        epochs = epochs
    )

### ANN with 2 layers of 60 and 20 neurons (with 0.1 Dropout Layer) (without scaling)

In [55]:
train_and_test_add_dropout(
    [60, 20],
    [0.1, 0],
    100,
    x, y
)

Epoch 1/100
268/268 [==============================] - 1s 3ms/step - loss: 0.4357 - accuracy: 0.7981 - val_loss: 0.3947 - val_accuracy: 0.8205
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3854 - accuracy: 0.8305 - val_loss: 0.3546 - val_accuracy: 0.8493
Epoch 3/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3815 - accuracy: 0.8307 - val_loss: 0.3775 - val_accuracy: 0.8318
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3822 - accuracy: 0.8265 - val_loss: 0.3540 - val_accuracy: 0.8484
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3758 - accuracy: 0.8320 - val_loss: 0.3481 - val_accuracy: 0.8485
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3716 - accuracy: 0.8335 - val_loss: 0.3482 - val_accuracy: 0.8496
Epoch 7/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3720 - accuracy: 0.8362 - val_loss: 0.3469 - val_accuracy: 0.8475

85.54% training accuracy with 85.30% validation accuracy

- accuracy dropped
- equivalent accuracy (training and validating)
- try adding another layer
- try increasing epochs

### ANN with 3 layers of 60, 30, 10 neurons (with 0.1 Dropout Layer) (without scaling)

In [56]:
train_and_test_add_dropout(
    [60, 30, 10],
    [0.1, 0, 0],
    100,
    x, y
)

Epoch 1/100
268/268 [==============================] - 2s 3ms/step - loss: 0.4563 - accuracy: 0.7850 - val_loss: 0.3588 - val_accuracy: 0.8436
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3897 - accuracy: 0.8260 - val_loss: 0.3501 - val_accuracy: 0.8485
Epoch 3/100
268/268 [==============================] - 1s 3ms/step - loss: 0.3888 - accuracy: 0.8261 - val_loss: 0.3506 - val_accuracy: 0.8446
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3825 - accuracy: 0.8313 - val_loss: 0.3496 - val_accuracy: 0.8488
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3826 - accuracy: 0.8295 - val_loss: 0.3483 - val_accuracy: 0.8471
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3782 - accuracy: 0.8323 - val_loss: 0.3369 - val_accuracy: 0.8537
Epoch 7/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3765 - accuracy: 0.8337 - val_loss: 0.3375 - val_accuracy: 0.8525

86.14% training accuracy with 85.56% validation accuracy

- accuracy increased
- equivalent accuracy (training and validating)
- try scaling

### ANN with 3 layers of 60, 30, 10 neurons (with 0.1 Dropout Layer) (with Standard scaling)

In [57]:
x_scaled = pd.DataFrame(StandardScaler().fit_transform(x), columns=x.columns)

In [60]:
train_and_test_add_dropout(
    [60, 30, 10],
    [0.1, 0, 0],
    100,
    x_scaled,
    y
)

Epoch 1/100
268/268 [==============================] - 2s 3ms/step - loss: 0.4472 - accuracy: 0.7926 - val_loss: 0.3580 - val_accuracy: 0.8452
Epoch 2/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3658 - accuracy: 0.8403 - val_loss: 0.3440 - val_accuracy: 0.8518
Epoch 3/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3522 - accuracy: 0.8481 - val_loss: 0.3422 - val_accuracy: 0.8519
Epoch 4/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3430 - accuracy: 0.8529 - val_loss: 0.3371 - val_accuracy: 0.8551
Epoch 5/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3380 - accuracy: 0.8535 - val_loss: 0.3344 - val_accuracy: 0.8566
Epoch 6/100
268/268 [==============================] - 1s 2ms/step - loss: 0.3329 - accuracy: 0.8567 - val_loss: 0.3334 - val_accuracy: 0.8586
Epoch 7/100
268/268 [==============================] - 1s 3ms/step - loss: 0.3294 - accuracy: 0.8578 - val_loss: 0.3308 - val_accuracy: 0.8590

90.11% training accuracy with 84.61% validation accuracy

- accuracy increased
- slightly overfitting
- try adding more dropout

### ANN with 3 layers of 60, 30, 10 neurons (with 0.2 Dropout Layer) (with Standard scaling)

In [72]:
train_and_test_add_dropout(
    [60, 40, 20],
    [0.1, 0.1, 0],
    200,
    x_scaled,
    y
)

Epoch 1/200
268/268 [==============================] - 2s 3ms/step - loss: 0.4390 - accuracy: 0.7958 - val_loss: 0.3596 - val_accuracy: 0.8443
Epoch 2/200
268/268 [==============================] - 1s 3ms/step - loss: 0.3702 - accuracy: 0.8379 - val_loss: 0.3467 - val_accuracy: 0.8519
Epoch 3/200
268/268 [==============================] - 1s 3ms/step - loss: 0.3553 - accuracy: 0.8453 - val_loss: 0.3387 - val_accuracy: 0.8562
Epoch 4/200
268/268 [==============================] - 1s 3ms/step - loss: 0.3489 - accuracy: 0.8475 - val_loss: 0.3368 - val_accuracy: 0.8561
Epoch 5/200
268/268 [==============================] - 1s 3ms/step - loss: 0.3452 - accuracy: 0.8497 - val_loss: 0.3402 - val_accuracy: 0.8552
Epoch 6/200
268/268 [==============================] - 1s 3ms/step - loss: 0.3432 - accuracy: 0.8520 - val_loss: 0.3352 - val_accuracy: 0.8590
Epoch 7/200
268/268 [==============================] - 1s 3ms/step - loss: 0.3401 - accuracy: 0.8538 - val_loss: 0.3337 - val_accuracy: 0.8580